### PR TITLE
sql/catalog: avoid fetching descriptors when fetching comments

### LINKED ISSUE #102613: sql: query performance regression in pg_catalog query that resolves database objects and their descriptions
This query is really slow and seems to be a regression in 23.1 compared to 22.2.8:
```
SELECT
  current_database(), n.nspname, relname, d.description
FROM
  pg_description AS d
  INNER JOIN pg_class AS c ON d.objoid = c.oid
  INNER JOIN pg_namespace AS n ON n.oid = c.relnamespace
WHERE
  d.objsubid = 0
  AND n.nspname
    NOT IN (
        'gp_toolkit':::STRING:::NAME,
        'information_schema':::STRING:::NAME,
        'pgagent':::STRING:::NAME
      )
  AND n.nspname NOT LIKE 'pg_%';
```

Explain:
```
root@localhost:26257/defaultdb> EXPLAIN ANALYZE (VERBOSE) SELECT
                             ->   current_database(), n.nspname, relname, d.description
                             -> FROM
                             ->   pg_description AS d
                             ->   INNER JOIN pg_class AS c ON d.objoid = c.oid
                             ->   INNER JOIN pg_namespace AS n ON n.oid = c.relnamespace
                             -> WHERE
                             ->   d.objsubid = 0
                             ->   AND n.nspname
                             ->     NOT IN (
                             ->         'gp_toolkit':::STRING:::NAME,
                             ->         'information_schema':::STRING:::NAME,
                             ->         'pgagent':::STRING:::NAME
                             ->       )
                             ->   AND n.nspname NOT LIKE 'pg_%';
                                                        info
--------------------------------------------------------------------------------------------------------------------
  planning time: 1ms
  execution time: 4.5s
  distribution: local
  vectorized: true
  maximum memory usage: 310 KiB
  network usage: 0 B (0 messages)
  sql cpu time: 4.5s
  regions: local

  • render
  │ columns: (current_database, nspname, relname, description)
  │ render current_database: 'defaultdb'
  │ render description: description
  │ render relname: relname
  │ render nspname: nspname
  │
  └── • hash join (inner)
      │ columns: (oid, nspname, objoid, objsubid, description, oid, relname, relnamespace)
      │ nodes: n1
      │ regions: local
      │ actual row count: 76
      │ vectorized batch count: 2
      │ estimated max memory allocated: 180 KiB
      │ estimated max sql temp disk usage: 0 B
      │ sql cpu time: 177µs
      │ estimated row count: 156 (missing stats)
      │ equality: (oid) = (relnamespace)
      │
      ├── • filter
      │   │ columns: (oid, nspname)
      │   │ nodes: n1
      │   │ regions: local
      │   │ actual row count: 2
      │   │ vectorized batch count: 2
      │   │ sql cpu time: 9µs
      │   │ estimated row count: 111 (missing stats)
      │   │ filter: (nspname NOT IN ('gp_toolkit', 'information_schema', 'pgagent')) AND (nspname NOT LIKE 'pg_%')
      │   │
      │   └── • virtual table
      │         columns: (oid, nspname)
      │         nodes: n1
      │         regions: local
      │         actual row count: 5
      │         vectorized batch count: 3
      │         sql cpu time: 69µs
      │         estimated row count: 1,000 (missing stats)
      │         table: pg_namespace@primary
      │
      └── • virtual table lookup join (inner)
          │ columns: (objoid, objsubid, description, oid, relname, relnamespace)
          │ nodes: n1
          │ regions: local
          │ actual row count: 294
          │ vectorized batch count: 9
          │ sql cpu time: 4.5s
          │ estimated row count: 99 (missing stats)
          │ table: pg_class@pg_class_oid_idx
          │ equality: (objoid) = (oid)
          │
          └── • filter
              │ columns: (objoid, objsubid, description)
              │ nodes: n1
              │ regions: local
              │ actual row count: 2,230
              │ vectorized batch count: 12
              │ sql cpu time: 14µs
              │ estimated row count: 10 (missing stats)
              │ filter: objsubid = 0
              │
              └── • virtual table
                    columns: (objoid, objsubid, description)
                    nodes: n1
                    regions: local
                    actual row count: 2,230
                    vectorized batch count: 12
                    sql cpu time: 2ms
                    estimated row count: 1,000 (missing stats)
                    table: pg_description@primary
(78 rows)

Time: 4.535s total (execution 4.535s / network 0.000s)
```

The slow part
```
      └── • virtual table lookup join (inner)
          │ columns: (objoid, objsubid, description, oid, relname, relnamespace)
          │ nodes: n1
          │ regions: local
          │ actual row count: 294
          │ vectorized batch count: 9
          │ sql cpu time: 4.5s
          │ estimated row count: 99 (missing stats)
          │ table: pg_class@pg_class_oid_idx
          │ equality: (objoid) = (oid)
```

However on 22.2.8 this is much faster
```
root@localhost:26257/defaultdb> EXPLAIN ANALYZE (VERBOSE) SELECT                                                                            current_database(), n.nspname, relname, d.description                                                                                   FROM                                                                                                                                        pg_description AS d                                                                                                                       INNER JOIN pg_class AS c ON d.objoid = c.oid                                                                                              INNER JOIN pg_namespace AS n ON n.oid = c.relnamespace                                                                                  WHERE                                                                                                                                       d.objsubid = 0                                                                                                                            AND n.nspname                                                                                                                               NOT IN (                                                                                                                                      'gp_toolkit':::STRING:::NAME,                                                                                                             'information_schema':::STRING:::NAME,                                                                                                     'pgagent':::STRING:::NAME                                                                                                               )                                                                                                                                     AND n.nspname NOT LIKE 'pg_%';
                                                        info
--------------------------------------------------------------------------------------------------------------------
  planning time: 834µs
  execution time: 18ms
  distribution: local
  vectorized: true
  maximum memory usage: 300 KiB
  network usage: 0 B (0 messages)
  regions: local

  • render
  │ columns: (current_database, nspname, relname, description)
  │ render current_database: 'defaultdb'
  │ render description: description
  │ render relname: relname
  │ render nspname: nspname
  │
  └── • hash join (inner)
      │ columns: (oid, nspname, objoid, objsubid, description, oid, relname, relnamespace)
      │ nodes: n1
      │ regions: local
      │ actual row count: 64
      │ vectorized batch count: 2
      │ estimated max memory allocated: 190 KiB
      │ estimated max sql temp disk usage: 0 B
      │ estimated row count: 156 (missing stats)
      │ equality: (oid) = (relnamespace)
      │
      ├── • filter
      │   │ columns: (oid, nspname)
      │   │ nodes: n1
      │   │ regions: local
      │   │ actual row count: 2
      │   │ vectorized batch count: 2
      │   │ estimated row count: 111 (missing stats)
      │   │ filter: (nspname NOT IN ('gp_toolkit', 'information_schema', 'pgagent')) AND (nspname NOT LIKE 'pg_%')
      │   │
      │   └── • virtual table
      │         columns: (oid, nspname)
      │         nodes: n1
      │         regions: local
      │         actual row count: 5
      │         vectorized batch count: 3
      │         estimated row count: 1,000 (missing stats)
      │         table: pg_namespace@primary
      │
      └── • virtual table lookup join (inner)
          │ columns: (objoid, objsubid, description, oid, relname, relnamespace)
          │ nodes: n1
          │ regions: local
          │ actual row count: 282
          │ vectorized batch count: 9
          │ estimated row count: 99 (missing stats)
          │ table: pg_class@pg_class_oid_idx
          │ equality: (objoid) = (oid)
          │
          └── • filter
              │ columns: (objoid, objsubid, description)
              │ nodes: n1
              │ regions: local
              │ actual row count: 282
              │ vectorized batch count: 9
              │ estimated row count: 10 (missing stats)
              │ filter: objsubid = 0
              │
              └── • virtual table
                    columns: (objoid, objsubid, description)
                    nodes: n1
                    regions: local
                    actual row count: 282
                    vectorized batch count: 9
                    estimated row count: 1,000 (missing stats)
                    table: pg_description@primary
(71 rows)


Time: 20ms total (execution 19ms / network 0ms)

root@localhost:26257/defaultdb>
```

[Statement Bundle](https://github.com/cockroachdb/cockroach/files/11354949/stmt-bundle-860564037657526273.zip)

Commit used to build cockroach: f8523ec35b2be41fb3a48e21a9d0b8ca52d75363

Jira issue: CRDB-27560

### PR DESCRIPTION
**Pull request by fqazi**
Previously, we refactored the code to fetch comments, descriptors, and zone config together in all cases. A side effect of this change was that the crdb_internal.kv_system_comments table was substantially slower for larger tables leading to big regressions. To address this, this patch adds a method of only fetching comments within the collections.

Informs: #102851
Fixes: #102613

Release note (bug fix): Optimize over-head of
pg_catalog.pg_description and pg_catalog.pg_shdescription, which can lead to performance regression relative to 22.2

### CODE REVIEW

File: pkg/sql/catalog/internal/catkv/catalog_reader_cached.go [hunk 1]
```diff
@@ -149,6 +150,26 @@ func (c *cachedCatalogReader) Reset(ctx context.Context) {
...
+) (nstree.Catalog, error) {
+	if c.hasScanAll || c.hasScanAllComments {
+		return c.cache.Catalog, nil
+	}
+	// Scan all catalog tables.
+	read, err := c.cr.ScanAllComments(ctx, txn)
+	if err != nil {
+		return nstree.Catalog{}, err
+	}
+	/*if err := c.ensure(ctx, read); err != nil {
```
Comment (chengxiong-ruan) [inline pkg/sql/catalog/internal/catkv/catalog_reader_cached.go:165]: nit: remove these

**Review by chengxiong-ruan — APPROVED** [review]
lgtm

**Comment by fqazi** [comment]
@chengxiong-ruan TFTR!

**Review by rafiss — CHANGES_REQUESTED** [review]
nice fix. can you add a test for this in rttanalysis?

**Review by rafiss — COMMENTED** [review]
we should be able to use the query that dikshant added in the ticket. it came from an ORM, so it can be in the ORMQueries benchmark test.
```
SELECT
  current_database(), n.nspname, relname, d.description
FROM
  pg_description AS d
  INNER JOIN pg_class AS c ON d.objoid = c.oid
  INNER JOIN pg_namespace AS n ON n.oid = c.relnamespace
WHERE
  d.objsubid = 0
  AND n.nspname
    NOT IN (
        'gp_toolkit':::STRING:::NAME,
        'information_schema':::STRING:::NAME,
        'pgagent':::STRING:::NAME
      )
  AND n.nspname NOT LIKE 'pg_%';
```

**Comment by fqazi** [comment]
> we should be able to use the query that dikshant added in the ticket. it came from an ORM, so it can be in the ORMQueries benchmark test.
> 
> ```
> SELECT
>   current_database(), n.nspname, relname, d.description
> FROM
>   pg_description AS d
>   INNER JOIN pg_class AS c ON d.objoid = c.oid
>   INNER JOIN pg_namespace AS n ON n.oid = c.relnamespace
> WHERE
>   d.objsubid = 0
>   AND n.nspname
>     NOT IN (
>         'gp_toolkit':::STRING:::NAME,
>         'information_schema':::STRING:::NAME,
>         'pgagent':::STRING:::NAME
>       )
>   AND n.nspname NOT LIKE 'pg_%';
> ```

@rafiss  weirdly rttanalysis chokes on this query. This one is bad; it takes 11 seconds on a  cockroach demo. I used the one from Django:

```
SELECT
                c.relname,
                CASE
                    WHEN c.relispartition THEN 'p'
                    WHEN c.relkind IN ('m', 'v') THEN 'v'
                    ELSE 't'
                END,
                obj_description(c.oid, 'pg_class')
            FROM pg_catalog.pg_class c
            LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
            WHERE c.relkind IN ('f', 'm', 'p', 'r', 'v')
                AND n.nspname NOT IN ('pg_catalog', 'pg_toast')
                AND pg_catalog.pg_table_is_visible(c.oid);
```

**Comment by fqazi** [comment]
Let me dig into the regression with the other one

**Pull request closed without merging**